Adds `portal.add_catalog_indexes` and `portal.add_catalog_metadata` methods (#578)
plone/plone.api: docs/portal.md

    assert vocabulary_name in vocabulary_names
```

(portal-add-catalog-indexes-example)=

## Add catalog indexes

> To add new indexes to the portal catalog if they don't already exist, use {meth}`api.portal.add_catalog_indexes`.

Reviewer: ```suggestion
To add indexes to the portal catalog, use {meth}`api.portal.add_catalog_indexes`.
This function returns a list of the names of the indexes that were added.

The following collection of code snippets demonstrate how to add indexes and either use default logging, to skip reindexing, or to use a customer logger.
```

```python
from plone import api

# Add new indexes with default logging
indexes = [
    ('custom_field', 'FieldIndex'),
    ('review_date', 'DateIndex')
]
api.portal.add_catalog_indexes(wanted_indexes=indexes)

# Add indexes but skip reindexing
api.portal.add_catalog_indexes(
    wanted_indexes=[('modified_by', 'FieldIndex')],
    reindex=False
)

# Add indexes with custom logger
import logging
custom_logger = logging.getLogger('my.package')
api.portal.add_catalog_indexes(
    wanted_indexes=[('creation_user', 'FieldIndex')],
    logger=custom_logger
)
```

% invisible-code-block: python
%
% # Verify the indexes were added to the catalog
% catalog = api.portal.get_tool('portal_catalog')
% self.assertIn('custom_field', catalog.indexes())
% self.assertIn('review_date', catalog.indexes())
% self.assertIn('modified_by', catalog.indexes())
% self.assertIn('creation_user', catalog.indexes())

> This function returns a list of the names of the indexes that were added.

Reviewer: Move above.

```suggestion
```

(portal-add-catalog-metadata-example)=

## Add catalog metadata columns

> To add new metadata columns to the portal catalog if they don't already exist, use {meth}`api.portal.add_catalog_metadata`.

Reviewer: ````suggestion
To add metadata columns to the portal catalog, use {meth}`api.portal.add_catalog_metadata`.
This function returns a list of the names of the columns that were added.

```{note}
Adding metadata columns only makes them available for storage.
You still need to reindex your content to populate the values.
```

The following collection of code snippets adds metadata columns with either the default logger or a custom logger.
````

```python
from plone import api

# Add new metadata columns with default logging
columns = ['custom_metadata', 'author_email']
api.portal.add_catalog_metadata(wanted_columns=columns)

# Add columns with custom logger
import logging
custom_logger = logging.getLogger('my.package')
api.portal.add_catalog_metadata(
    wanted_columns=['publication_date'],
    logger=custom_logger
)
```

% invisible-code-block: python
%
% # Verify the columns were added to the catalog
% catalog = api.portal.get_tool('portal_catalog')
% self.assertIn('custom_metadata', catalog.schema())
% self.assertIn('author_email', catalog.schema())
% self.assertIn('publication_date', catalog.schema())

> This function returns a list of the names of the columns that were added. Note that adding metadata columns only makes them available for storage - you still need to reindex your content to populate the values.

Reviewer: Moved above, and created an admonition for the note.

```suggestion
```

    assert vocabulary_name in vocabulary_names
```

(portal-add-catalog-indexes-example)=

## Add catalog indexes

To add indexes to the portal catalog, use {meth}`api.portal.add_catalog_indexes`.
This function returns a list of the names of the indexes that were added.

The following collection of code snippets demonstrate how to add indexes and either use default logging, to skip reindexing, or to use a customer logger.

```python
from plone import api

# Add a single field index
api.portal.add_catalog_indexes([('my_custom_field', 'FieldIndex')])

# Add multiple indexes with different types
indexes_to_add = [
    ('text_content', 'ZCTextIndex'),
    ('tags', 'KeywordIndex')
]
api.portal.add_catalog_indexes(indexes_to_add)

# Add indexes without reindexing
api.portal.add_catalog_indexes([('quick_field', 'FieldIndex')], reindex=False)

> ```

Reviewer: Why did you remove the doctest that you added earlier?

```suggestion

% invisible-code-block: python
%
% # Verify the indexes were added to the catalog
% catalog = api.portal.get_tool('portal_catalog')
% self.assertIn('custom_field', catalog.indexes())
% self.assertIn('review_date', catalog.indexes())
% self.assertIn('modified_by', catalog.indexes())
% self.assertIn('creation_user', catalog.indexes())

```

Author: i actually wrote few things from scratch, during adding `ZCTextIndex` handling. Might have been omitted doing it! 😅

### ZCTextIndex Special Handling

When adding a `ZCTextIndex`, the function automatically applies additional parameters:

> - `lexicon_id`: Set to 'plone_lexicon' by default
> - `index_type`: Set to 'Okapi BM25 Rank'

Reviewer: These are inline literals.

```suggestion

- `lexicon_id`: Set to `plone_lexicon` by default
- `index_type`: Set to `Okapi BM25 Rank`
```

    assert vocabulary_name in vocabulary_names
```

(portal-add-catalog-indexes-example)=

## Add catalog indexes

To add indexes to the portal catalog, use {meth}`api.portal.add_catalog_indexes`.
This function returns a list of the names of the indexes that were added.

The following collection of code snippets demonstrate how to add indexes and either use default logging, to skip reindexing, or to use a customer logger.

```python
from plone import api

# Add a single field index
api.portal.add_catalog_indexes([('my_custom_field', 'FieldIndex')])

# Add multiple indexes with different types
indexes_to_add = [
    ('text_content', 'ZCTextIndex'),
    ('tags', 'KeywordIndex')
]
api.portal.add_catalog_indexes(indexes_to_add)

# Add indexes without reindexing
api.portal.add_catalog_indexes([('quick_field', 'FieldIndex')], reindex=False)
```

> % invisible-code-block: python

Reviewer: ```suggestion

% invisible-code-block: python
```
